adding SHWFS class and corresponding demo

remorgan123 opened the pull request:
Hi,

Here's some code Ewan Douglas (@douglase )/I have developed to model a SHWFS through POPPY. It essentially subsamples a wavefront and calculates each spot PSF separately, then tiles it back together to return the full sensor measurement. There are also methods to calculate the centroids of lenslet spots and reconstruct the wavefront as a part of the class. There is a demo notebook in "notebooks/" to show how the class works. Currently, the SHWFS is in the "subapertures.py" file within POPPY, and it works directly on wavefront objects. Since it's a combination of apertures/optical elements/a detector I'm not sure how best to incorporate it into an Optical System class other than retrieving the wavefront at the end of the optical train and measuring it with a SHWFS object. Let me know if you have any suggestions for this or think it would be more appropriate as it's own repository separate from the main POPPY source code!

Thanks,
Rachel

douglase commented:
(this was also anticipated as a 1st step to modeling other imaging slicing instruments, such as integral field spectrographs, hence the general subapertures class).

mperrin commented on poppy/sub_sampled_optics.py:
```diff
@@ -0,0 +1,429 @@
+        
+        Parameters
+        ----------
+        
```
[change] Doc string is incomplete, missing descriptions of parameters including optic_array and overwrite_input_wavefront

mperrin commented on poppy/sub_sampled_optics.py:
```diff
@@ -0,0 +1,429 @@
+            self.optic_array=optic_array
+            self.crosstalk=crosstalk
+            if crosstalk:
+                raise ValueError("CROSS TALK NOT IMPLEMENTED <YET>")
```
[optional] This could raise a `NotImplementedError` instead

mperrin commented on poppy/sub_sampled_optics.py:
```diff
@@ -0,0 +1,429 @@
+                self.detector=detector
+            self.optical_system = optical_system
+            if  optical_system is not None:
+                    raise ValueError("complete optical system after wavelets are not implemented yet")
```
Same re `NotImplementedError`

mperrin commented on poppy/sub_sampled_optics.py:
```diff
@@ -0,0 +1,429 @@
+                
+        def sample_wf(self, wf):
+            '''
+            subsamples wavefront optic in order to mupliply by optic array
```
minor: spelling `mupliply`

mperrin commented on poppy/sub_sampled_optics.py:
```diff
@@ -0,0 +1,429 @@
+                    
+                    self.c = wf.wavefront.shape[0]/2*u.pix #center of array
+                    c=self.c
+                    w=self._w
```
The very short variable names here (`c`, `w`) are more cryptic than I would prefer. Spelling these out as `width` and `center` can make the code more maintainable in the long run. 

Also, why duplicate `c=self.c` and so forth? Seems unnecessary.

mperrin commented on poppy/sub_sampled_optics.py:
```diff
@@ -0,0 +1,429 @@
+        
+        #return a composite wavefront if an array of output wavefronts was generated
+        def get_wavefront_array(self):
+            """
```
[change] Blank doc string.

mperrin commented on poppy/sub_sampled_optics.py:
```diff
@@ -0,0 +1,429 @@
+                        wf.wavefront[lower_x:upper_x,lower_y:upper_y] = sub_wf.wavefront 
+            return 
+
+        def reconstruct_wavefront(self, flat_centroid_list ):
```
Should wavefront reconstruction be in the SH_WFS class instead of the base Subapertures class?

mperrin commented on poppy/sub_sampled_optics.py:
```diff
@@ -0,0 +1,429 @@
+import logging
+_log = logging.getLogger('poppy')
+
+class subapertures(poppy.OpticalElement):
```
[change] Please rename to `Subapertures`, as per PEP8 style guide recommendation of class names to start with an initial capital letter.

mperrin commented on poppy/sub_sampled_optics.py:
```diff
@@ -0,0 +1,429 @@
+            return wf_reconstruction
+
+
+class SH_WFS(subapertures):
```
Perhaps rename `ShackHartmannWavefrontSensor` to be more explicit.

mperrin commented on poppy/sub_sampled_optics.py:
```diff
@@ -0,0 +1,429 @@
+        pixel pitch
+        number of lenslets per axis (assuming square lenslet array with n_lenslets**2 lenslets total)
+        circular aperture (True means supapertures are circular, False means they are square)
+        detector: Detector object defining pixel pitch and plate scale
```
[change] Please update doc string for consistency with the numpy docstring standard. See https://numpydoc.readthedocs.io/en/latest/format.html

mperrin commented on poppy/sub_sampled_optics.py:
```diff
@@ -0,0 +1,429 @@
+        optic_array = np.array([[aperture,
+                                              aperture],
+                                    [aperture,
+                                    aperture]])
```
minor: clean up indentation

mperrin commented on poppy/sub_sampled_optics.py:
```diff
@@ -0,0 +1,429 @@
+         HDU.header['name']=self.name
+         HDU.header['SH_pitch'] = self.lenslet_pitch.to(u.m).value
+         HDU.header['SH_fl'] = self.lenslet_fl.to(u.m).value
+         HDU.header['DETpitch'] = self.pixel_pitch.to(u.m/u.pix).value
```
[optional] These can all be assigned as tuples, to include descriptive comments in the FITS header as well as the keyword values.

mperrin commented on poppy/sub_sampled_optics.py:
```diff
@@ -0,0 +1,429 @@
+    def calculate_centroid_requirement(self, min_WFE):
+        """
+          centroid 
+        """
```
[change] incomplete doc string

mperrin commented on poppy/sub_sampled_optics.py:
```diff
@@ -0,0 +1,429 @@
+        return  1.0/self.lenslet_fl.to(u.m)*self.lenslet_pitch.to(u.m)**2/2.0
+
+#helper functions for wavefront reconstruction:
+def zonalReconstruction(x,y,ds):
```
[change] rename to `zonal_reconstruction` as per PEP8 style guide recommendation that function names should be lowercase, with words separated by underscores.

mperrin commented on poppy/sub_sampled_optics.py:
```diff
@@ -0,0 +1,429 @@
+            E[m*(n-1)+i*(m-1)+j, i+j*n] = -1 
+            E[m*(n-1)+i*(m-1)+j, i+(j+1)*n] = 1
+    return E
+
```
These are pretty cryptic. Doc strings strongly recommended. Are there use cases for calling these functions independently outside of the zonal_reconstruction function?  If not, these definitions could either be moved inside the zonal_reconstruction function, or have initial underscores added to the names to indicate they're internal functions not for general purpose use.

remorgan123 commented:
Thanks Marshall! I'll work on these suggestions in the next week or so.
Thanks for the review!
Best,
Rachel

On Mon, Mar 2, 2020 at 5:10 PM Marshall Perrin <notifications@github.com>
wrote:

> *@mperrin* requested changes on this pull request.
>
> Hi @remorgan123 <https://github.com/remorgan123> and @douglase
> <https://github.com/douglase> - Thanks very much for this PR! I've tested
> the code out and it's working as it should. I'll definitely be glad to
> include this all into poppy, but I think a few parts need a little bit of
> revision before we do so.
>
> I've provided some suggestions in the attached code review, most of which
> should be relatively straightforward changes (completing some doc strings,
> renaming a few items for consistency with the Python style guide, etc).
>
> The demo notebook is a great start at showing the functionality. However
> right now it lacks any explanatory text about what's going on. A few brief
> sentences here and there would go a long way in helping a new user
> understand what's going on, and how to make use of this functionality.
>
> Let me know if you have questions about any of the suggestions and
> comments. Cheers, and thanks for contributing this!
> ------------------------------
>
> In poppy/sub_sampled_optics.py
> <https://github.com/spacetelescope/poppy/pull/346#discussion_r386667616>:
>
> > +
> +        subaperture(ARRAY OF OPTICS)
> +        #initialize this new class, where the array of optics define the subapertures (e.g. lenslets)
> +
> +        subapertures.sample_wf(wf) #this function takes the wavefront and subsamples it by the area of each optic
> +        subapertures.get_wavefront_array() #returns an array of input sub-wavefronts multipled by subaperture optics
> +        subapertures.get_psfs() #fraunhofer or fresnel propagation of each pupil  to the image/ waist
> +        image=subapertures.get_composite_wavefont() # returns wavefront of image plane of all the spots put back together
> +        subapertures.opd #single array made up of subapertures
> +        subapertures.amplitude  #single array made up of subapertures
> +        subapertures.getphasor #returns propagated array of spots from get_composite_wavefont
> +
> +
> +        Parameters
> +        ----------
> +
>
> [change] Doc string is incomplete, missing descriptions of parameters
> including optic_array and overwrite_input_wavefront
> ------------------------------
>
> In poppy/sub_sampled_optics.py
> <https://github.com/spacetelescope/poppy/pull/346#discussion_r386667868>:
>
> > +                                     CircularAperture(radius=2.,planetype=PlaneType.pupil)]]),
> +                     crosstalk=False,
> +                     x_y_offset=(0,0),
> +                     detector=None,
> +                     overwrite_inputwavefront=False,
> +                     display_intermediates=False,
> +                     optical_system=None,
> +                     subwf_oversample=3,
> +                    **kwargs):
> +            dimensions = optic_array.shape
> +            self.n_apertures = dimensions[0]*dimensions[1]
> +            self.subwf_oversample = subwf_oversample
> +            self.optic_array=optic_array
> +            self.crosstalk=crosstalk
> +            if crosstalk:
> +                raise ValueError("CROSS TALK NOT IMPLEMENTED <YET>")
>
> [optional] This could raise a NotImplementedError instead
> ------------------------------
>
> In poppy/sub_sampled_optics.py
> <https://github.com/spacetelescope/poppy/pull/346#discussion_r386668119>:
>
> > +            self.optic_array=optic_array
> +            self.crosstalk=crosstalk
> +            if crosstalk:
> +                raise ValueError("CROSS TALK NOT IMPLEMENTED <YET>")
> +            self.x_y_offset=x_y_offset
> +            self.amplitude = np.asarray([1.])
> +            self.opd = np.asarray([0.])
> +            self.input_wavefront = None
> +            self.output_wavefront = None
> +            if detector == None:
> +                self.detector = Detector(0.01,fov_pixels=128)
> +            else:
> +                self.detector=detector
> +            self.optical_system = optical_system
> +            if  optical_system is not None:
> +                    raise ValueError("complete optical system after wavelets are not implemented yet")
>
> Same re NotImplementedError
> ------------------------------
>
> In poppy/sub_sampled_optics.py
> <https://github.com/spacetelescope/poppy/pull/346#discussion_r386668351>:
>
> > +            self.y_apertures=self.optic_array.shape[1]
> +            if self.x_apertures != self.y_apertures:
> +                raise ValueError("A square array of subapertures is currently required.")
> +            #initialize array of subsampled output wavefronts:
> +            self.wf_array = np.empty(self.optic_array.shape,dtype=np.object_)
> +            self.overwrite_inputwavefront = overwrite_inputwavefront
> +            self.display_intermediates = display_intermediates
> +            self._propagated_flag = False #can't have propagated when initializing
> +            self._centroided_flag = False #can't have centroids without propagating
> +
> +            OpticalElement.__init__(self, **kwargs)
> +            return
> +
> +        def sample_wf(self, wf):
> +            '''
> +            subsamples wavefront optic in order to mupliply by optic array
>
> minor: spelling mupliply
> ------------------------------
>
> In poppy/sub_sampled_optics.py
> <https://github.com/spacetelescope/poppy/pull/346#discussion_r386669051>:
>
> > +                    self._w = opt.pupil_diam.to(u.m)/wf.pixelscale.to(u.m/u.pix) #subaperture width in pixels
> +                    #the generated number of subapertures might not match the input wavefront dimensions
> +
> +                    #want to center the subapertures on the incoming wavefront
> +
> +                    self.c = wf.wavefront.shape[0]/2*u.pix #center of array
> +                    c=self.c
> +                    w=self._w
>
> The very short variable names here (c, w) are more cryptic than I would
> prefer. Spelling these out as width and center can make the code more
> maintainable in the long run.
>
> Also, why duplicate c=self.c and so forth? Seems unnecessary.
> ------------------------------
>
> In poppy/sub_sampled_optics.py
> <https://github.com/spacetelescope/poppy/pull/346#discussion_r386669345>:
>
> > +                    # multiply by subaperture optic:
> +                    self.wf_array[i][j] = sub_wf*opt
> +
> +                    if self.display_intermediates:
> +                        plt.figure()
> +                        self.wf_array[i][j].display()
> +            return
> +
> +        @property
> +        def subaperture_width(self):
> +            # returns width in angular units
> +            return self._w*self.input_wavefront.pixelscale
> +
> +        #return a composite wavefront if an array of output wavefronts was generated
> +        def get_wavefront_array(self):
> +            """
>
> [change] Blank doc string.
> ------------------------------
>
> In poppy/sub_sampled_optics.py
> <https://github.com/spacetelescope/poppy/pull/346#discussion_r386669940>:
>
> > +            for i in range(self.x_apertures):
> +                 for j in  range(self.y_apertures):
> +                    sub_wf = self.wf_array[i][j] #get an subaperture wavefront
> +                    lower_x = int((c + self._w*(i)  - self._w*self.x_apertures/2).value)
> +                    lower_y = int((c + self._w*(j)  - self._w*self.y_apertures/2).value)
> +                    upper_x = int((c + self._w*(i+1)  - self._w*self.x_apertures/2).value)
> +                    upper_y = int((c + self._w*(j+1)  - self._w*self.y_apertures/2).value)
> +                    #check for padding
> +
> +                    if sub_wf == None:
> +                        wf.wavefront[lower_x:upper_x,lower_y:upper_y] = np.nan
> +                    else:
> +                        wf.wavefront[lower_x:upper_x,lower_y:upper_y] = sub_wf.wavefront
> +            return
> +
> +        def reconstruct_wavefront(self, flat_centroid_list ):
>
> Should wavefront reconstruction be in the SH_WFS class instead of the base
> Subapertures class?
> ------------------------------
>
> In poppy/sub_sampled_optics.py
> <https://github.com/spacetelescope/poppy/pull/346#discussion_r386670233>:
>
> > +Developed by Rachel Morgan, 2019-2020
> +"""
> +
> +import numpy as np
> +import poppy
> +import matplotlib.pyplot as plt
> +import astropy.units as u
> +import astropy.io.fits as fits
> +from .poppy_core import OpticalElement, Detector, Wavefront, PlaneType, _PUPIL, _IMAGE, _RADIANStoARCSEC
> +from .optics import CircularAperture
> +from .utils import measure_centroid
> +
> +import logging
> +_log = logging.getLogger('poppy')
> +
> +class subapertures(poppy.OpticalElement):
>
> [change] Please rename to Subapertures, as per PEP8 style guide
> recommendation of class names to start with an initial capital letter.
> ------------------------------
>
> In poppy/sub_sampled_optics.py
> <https://github.com/spacetelescope/poppy/pull/346#discussion_r386670493>:
>
> > +                _log.warn("Trying to reconstruct wavefront without having propagated to detector. run sample_wf on wavefront and get_psfs first.")
> +            self.get_centroids()
> +            if flat_centroid_list.all()!=None:
> +                y_cen, x_cen = self.centroid_list
> +                y_cen_flat, x_cen_flat = flat_centroid_list
> +                x_off = np.subtract(x_cen, x_cen_flat)
> +                y_off = np.subtract(y_cen, y_cen_flat)
> +            else:
> +                y_off, x_off = self.centroid_list
> +            scale = self.pixel_pitch.to(u.m)/self.lenslet_fl.to(u.m)*u.rad/self.detector.oversample
> +            wf_reconstruction = zonalReconstruction(x_off*scale, y_off*scale, self.lenslet_pitch)
> +
> +            return wf_reconstruction
> +
> +
> +class SH_WFS(subapertures):
>
> Perhaps rename ShackHartmannWavefrontSensor to be more explicit.
> ------------------------------
>
> In poppy/sub_sampled_optics.py
> <https://github.com/spacetelescope/poppy/pull/346#discussion_r386671570>:
>
> > +        Shack-Hartmann Wavefront Sensor Class
> +        wrapper for subapertures class
> +        Parameters:
> +        lenslet pitch
> +        lenslet focal length
> +        pixel pitch
> +        number of lenslets per axis (assuming square lenslet array with n_lenslets**2 lenslets total)
> +        circular aperture (True means supapertures are circular, False means they are square)
> +        detector: Detector object defining pixel pitch and plate scale
>
> [change] Please update doc string for consistency with the numpy docstring
> standard. See https://numpydoc.readthedocs.io/en/latest/format.html
> ------------------------------
>
> In poppy/sub_sampled_optics.py
> <https://github.com/spacetelescope/poppy/pull/346#discussion_r386671701>:
>
> > +        optic_array = np.array([[aperture,
> +                                              aperture],
> +                                    [aperture,
> +                                    aperture]])
>
> minor: clean up indentation
> ------------------------------
>
> In poppy/sub_sampled_optics.py
> <https://github.com/spacetelescope/poppy/pull/346#discussion_r386672120>:
>
> > +         HDU.header['SH_units']='meters'
> +         HDU.header['name']=self.name
> +         HDU.header['SH_pitch'] = self.lenslet_pitch.to(u.m).value
> +         HDU.header['SH_fl'] = self.lenslet_fl.to(u.m).value
> +         HDU.header['DETpitch'] = self.pixel_pitch.to(u.m/u.pix).value
>
> [optional] These can all be assigned as tuples, to include descriptive
> comments in the FITS header as well as the keyword values.
> ------------------------------
>
> In poppy/sub_sampled_optics.py
> <https://github.com/spacetelescope/poppy/pull/346#discussion_r386672268>:
>
> > +        """
> +          centroid
> +        """
>
> [change] incomplete doc string
> ------------------------------
>
> In poppy/sub_sampled_optics.py
> <https://github.com/spacetelescope/poppy/pull/346#discussion_r386672896>:
>
> > +        return centroid
> +    @property
> +    def pix_lenslet(self):
> +        return self.lenslet_pitch.to(u.m)/self.pixel_pitch.to(u.m)
> +    @property
> +    def max_WFE(self):
> +        """
> +        returns the maximum wavefront error detectable for ideal lenslet before the spot crosses
> +        into the neighboring lenslet
> +
> +        """
> +        _log.warn("This max wavefront error ignores lenslet aberrations")
> +        return  1.0/self.lenslet_fl.to(u.m)*self.lenslet_pitch.to(u.m)**2/2.0
> +
> +#helper functions for wavefront reconstruction:
> +def zonalReconstruction(x,y,ds):
>
> [change] rename to zonal_reconstruction as per PEP8 style guide
> recommendation that function names should be lowercase, with words
> separated by underscores.
> ------------------------------
>
> In poppy/sub_sampled_optics.py
> <https://github.com/spacetelescope/poppy/pull/346#discussion_r386673911>:
>
> > +def getC(m,n):
> +    C = np.matrix(np.zeros([(m-1)*n+(n-1)*m, 2*n*m]))
> +
> +    for i in range(m):
> +        for j in range(n-1) :
> +                C[i*(n-1)+j, i*n+j]=0.5
> +                C[i*(n-1)+j, i*n+j+1]=0.5
> +    for i in range(n):
> +        for j in range(m-1):
> +                C[m*(n-1)+i*(m-1)+j, n*(m+j)+i]=0.5
> +                C[m*(n-1)+i*(m-1)+j, n*(m+j+1)+i]=0.5
> +    return C
> +
> +
> +def getE(m,n):
> +    E=np.matrix(np.zeros([(m-1)*n+(n-1)*m, n*m]))
> +    for i in range(m):
> +        for j in range(n-1):
> +            E[i*(n-1)+j, i*n+j] = -1
> +            E[i*(n-1)+j, i*n+j+1] = 1
> +    for i in range(n):
> +        for j in range(m-1):
> +            E[m*(n-1)+i*(m-1)+j, i+j*n] = -1
> +            E[m*(n-1)+i*(m-1)+j, i+(j+1)*n] = 1
> +    return E
> +
>
> These are pretty cryptic. Doc strings strongly recommended. Are there use
> cases for calling these functions independently outside of the
> zonal_reconstruction function? If not, these definitions could either be
> moved inside the zonal_reconstruction function, or have initial underscores
> added to the names to indicate they're internal functions not for general
> purpose use.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spacetelescope/poppy/pull/346?email_source=notifications&email_token=AOK5RQK2S4KM44T5ZMFU6ELRFQVEDA5CNFSM4KKLKFBKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCXUFU2A#pullrequestreview-367549032>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOK5RQMQ6XYII5UTMF3SYVTRFQVEDANCNFSM4KKLKFBA>
> .
>

remorgan123 commented:
Hi Marshall, 

I just updated the docstrings and example notebook and tried to make the zonal reconstruction methods easier to read/understand. Please let me know if you have any more suggestions! Thanks for your review!

Best,
Rachel

remorgan123 commented:
Hi Marshall,

Thanks for the suggestions! I just pushed some more changes:

You're right - the SHWFS Demo.ipynb should be deleted. I deleted it from my
branch, thanks!

The max iterations warning is coming from the fwcentroid function, I'm not
sure why it's happening. It's not necessarily expected but the returned
results have all been reasonable, maybe the default threshold is a bit too
low for this application? It looks like you wrote that function, let me
know if you have any intuition/suggestions for it :)

I implemented a factor_of_two option in dms.py as you suggested, let me
know what you think!

I added your plotting code at the end of the demo, thanks for writing it!

Best,
Rachel

On Wed, Mar 18, 2020 at 11:46 AM Marshall Perrin <notifications@github.com>
wrote:

> *@mperrin* requested changes on this pull request.
>
> Hi @remorgan123 <https://github.com/remorgan123> - Thanks much. A few
> more suggestions & requests:
>
>    -
>
>    I see this currently contains both "SHWFS Demo.ipynb" and "Shack
>    Hartmann Wavefront Sensor Demo.ipynb". It looks like that's redundant and
>    the "SHWFS" one should be deleted - is that correct?
>    -
>
>    When running the example notebook, in cell 6 with the zonal
>    reconstruction it outputs some errors "Algorithm terminated at max
>    iterations without convergence.". Is that expected, and if so why? If it is
>    expected, could you add a sentence or two to the notebook to explain why
>    it's ok, please. (It's totally fine if the code is not 100% bulletproof and
>    robust - I'm just trying to understand how a user of this code might be
>    able to determine if something is a real concern or not).
>    -
>
>    I already commented separately that I would like to make any factor of
>    2 change for surface-to-OPD in the DM code optional, to avoid introducing
>    that change into existing working application code that doesn't expect it.
>    (I saw that in the version of this PR at mperrin/poppy#267
>    <https://github.com/mperrin/poppy/pull/267>, but don't see it here
>    now...)
>    -
>
>    And lastly, here's a suggestion for the last cell of the example
>    notebook. We can adjust the plot parameters somewhat and also include a
>    comparison plot of the input wavefront, to show that the reconstructed
>    wavefront is correctly sampling and reconstructing it. Try like this:
>
> #display result:
> plt.figure()
> plt.imshow(reconstruction.value, cmap=poppy.conf.cmap_diverging, vmin=-1, vmax=1)
> plt.colorbar(label = "Wavefront error, " + str(reconstruction.unit))
> plt.xlabel("x lenslet")
> plt.ylabel("y lenslet")
> plt.title("Reconstructed Wavefront")
>
>
> plt.figure()
> dm.display(opd_vmax=1e-6,
>            colorbar_orientation='vertical',
>            title='Input Wavefront (Truth)')
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spacetelescope/poppy/pull/346#pullrequestreview-376993562>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOK5RQN5K5N73WMCNMHK77LRIDUEZANCNFSM4KKLKFBA>
> .
>

mperrin commented:
@remorgan123 - I realized in all the chaos and disruptions of the last few months I hadn't yet merged this PR. I can go ahead and merge it now, but wanted to double check and make sure you agree it's ready to go and aren't planning any additional changes at this time. Thanks for contributing this! It's a nice piece of functionality.

remorgan123 commented:
Hi Marshall,
I’m good to go! Thanks for your helpful reviews!
Best,
Rachel

On Tue, May 5, 2020 at 9:53 AM Marshall Perrin <notifications@github.com>
wrote:

> @remorgan123 <https://github.com/remorgan123> - I realized in all the
> chaos and disruptions of the last few months I hadn't yet merged this PR. I
> can go ahead and merge it now, but wanted to double check and make sure you
> agree it's ready to go and aren't planning any additional changes at this
> time. Thanks for contributing this! It's a nice piece of functionality.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spacetelescope/poppy/pull/346#issuecomment-624068957>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOK5RQOAPRIWPYOR2YIL4VDRQAK4LANCNFSM4KKLKFBA>
> .
>
